[stdlib_math] Add function `diff` (#605)
fortran-lang/stdlib: doc/specs/stdlib_math.md

    
end program demo_math_all_close
```

### `diff`

#### Description

> Computes differences between adjacent elements of the input array.

Reviewer: ```suggestion
Computes differences between adjacent elements of an array.
```

#### Syntax

> ```fortran
> !> For rank-1 array
> Y = [[stdlib_math(module):diff(interface)]](X [, n])
> !> and for rank-2 array
> Y = [[stdlib_math(module):diff(interface)]](X [, n, dim])
> ```

Reviewer: ````suggestion
!> For rank-1 array
```fortran
Y = [[stdlib_math(module):diff(interface)]](X [, n])
```
and for a rank-2 array
```fortran
Y = [[stdlib_math(module):diff(interface)]](X [, n, dim])
```
````

#### Status

Experimental.

#### Class

Pure function.

#### Arguments

Note: If the value of `dim` is not equal to `1` or `2`,

> `1` will be used by the internal process of `diff`. (Not recommended)

Reviewer: I suggest to mention that after defining all arguments. Because, at this stage, the reader doesn't know what `dim` is (and for what it is used

`X`: Shall be a `real/integer` and `rank-1/rank-2` array.
This argument is `intent(in)`.

`n`: Shall be an `integer` scalar.
This argument is `intent(in)` and `optional`, which is `1` by default.
It represents to calculate the n-th order difference.

`dim`: Shall be an `integer` scalar.
This argument is `intent(in)` and `optional`, which is `1` by default.
It represents to calculate the difference along which dimension.

#### Result value

> Note: That `Y` has one fewer element than `X`.
> 
> Returns a `real/integer` and `rank-1/rank-2` array.

Reviewer: ```suggestion
Returns a `real/integer` and `rank-1/rank-2` array. The size of `Y` is the size of `X` minus 1.
```

    
end program demo_math_all_close
```

### `diff`

#### Description

Computes differences between adjacent elements of an array.

#### Syntax

For rank-1 array
```fortran
y = [[stdlib_math(module):diff(interface)]](x [, n, prepend, append])
```
and for rank-2 array
```fortran
y = [[stdlib_math(module):diff(interface)]](x [, n, dim, prepend, append])
```

#### Status

Experimental.

#### Class

Pure function.

#### Arguments

Note: The `x`, `prepend` and `append` arguments must have same `type`, `kind` and `rank`.

`x`: Shall be a `real/integer` and `rank-1/rank-2` array.
This argument is `intent(in)`.

`n`: Shall be an `integer` scalar.
This argument is `intent(in)` and `optional`, which is `1` by default.
It represents to calculate the n-th order difference.

`dim`: Shall be an `integer` scalar.
This argument is `intent(in)` and `optional`, which is `1` by default.

> It represents to calculate the difference along which dimension.

Reviewer: This sentence sounds strange to me. What do you mean?

Author: [numpy.diff](https://numpy.org/doc/stable/reference/generated/numpy.diff.html):
> `axis`: `int`, `optional`
The axis along which the difference is taken, default is the last axis.

[matlab.diff](https://ww2.mathworks.cn/help/matlab/ref/diff.html?lang=en):
> `Y = diff(X,n,dim)` is the nth difference calculated along the dimension specified by `dim`. The dim input is a positive integer scalar.

Reviewer 2: Perhaps change the last sentence to: `It represents the dimension along which the difference is calculated.`

    
end program demo_math_all_close
```

### `diff`

#### Description

Computes differences between adjacent elements of an array.

#### Syntax

For a rank-1 array
```fortran
y = [[stdlib_math(module):diff(interface)]](x [, n, prepend, append])
```
and for a rank-2 array
```fortran
y = [[stdlib_math(module):diff(interface)]](x [, n, dim, prepend, append])
```

#### Status

Experimental.

#### Class

Pure function.

#### Arguments

> `x`: Shall be a `real/integer` and `rank-1/rank-2` array.

Reviewer 3: ```suggestion
`x`: The array to take a difference of.
Shall be a `real/integer` and `rank-1/rank-2` array.
```

This argument is `intent(in)`.

> `n`: Shall be an `integer` scalar.
> This argument is `intent(in)` and `optional`, which is `1` by default.
> It represents to calculate the n-th order difference.

Reviewer 3: ```suggestion
`n`: The order of finite differencing to use.
Shall be an `integer` scalar.
This argument is `intent(in)` and `optional` and has value `1` by default.
```

> `dim`: Shall be an `integer` scalar.
> This argument is `intent(in)` and `optional`, which is `1` by default.
> It gives the dimension of the input array along which the difference is calculated, between `1` and `rank(x)`.

Reviewer 3: ```suggestion
`dim`: The dimension of the input array along which to calculate the difference.
Its value must be between `1` and `rank(x)`.
Shall be an `integer` scalar.
This argument is `intent(in)` and `optional` and has a value of `1` by default.
```

`prepend`: Shall be a `real/integer` and `rank-1/rank-2` array, which is no value by default.
This argument is `intent(in)` and `optional`.

`append`: Shall be a `real/integer` and `rank-1/rank-2` array, which is no value by default.

> This argument is `intent(in)` and `optional`.

Reviewer 3: Can you add 1-sentence description of these arguments (what do they mean)?

Note: 
- The `x`, `prepend` and `append` arguments must have the same `type`, `kind` and `rank`.

> - If the value of `n` is less than or equal to `0`, the return value of `y` is `x`. (Not recommended)

Reviewer 3: Do you mean `diff` instead of `y`?

- If the value of `dim` is not equal to `1` or `2`,

> `1` will be used by the internal process of `diff`. (Not recommended)

Reviewer 3: This only matters in the case of a rank-2 array, right?

If this is not recommended, why is `dim=1` the default value? I think the default value should be the recommended one.

Author: ```suggestion
- If the value of `dim` is not equal to `1` or `2` (which is not recommended),
`1` will be used by the internal process of `diff`.
```
How about changing it to this?

#### Result value

Returns a `real/integer` and `rank-1/rank-2` array.

> When both `prepend` and `append` are not present, the result `y` has one fewer element than `x` alongside the dimension `dim`.

Reviewer 3: It would be useful to here also state the meaning of the resulting value (the finite difference of the input array).

#### Example

```fortran
program demo_diff

    use stdlib_math, only: diff
    implicit none
    
    integer :: i(7) = [1, 1, 2, 3, 5, 8, 13]
    real    :: x(6) = [0, 5, 15, 30, 50, 75]
    integer :: A(3, 3) = reshape([1, 7, 17, 3, 11, 19, 5, 13, 23], [3, 3])
    integer :: Y(3, 2)
    
    print *, diff(i)        !! [0, 1, 1, 2, 3, 5]
    print *, diff(x, 2)     !! [5.0, 5.0, 5.0, 5.0]
    
    Y = diff(A, n=1, dim=2)
    print *, Y(1, :)        !! [2, 2]
    print *, Y(2, :)        !! [4, 2]
    print *, Y(3, :)        !! [2, 4]

    print *, diff(i, prepend=[0]) !! [1, 0, 1, 1, 2, 3, 5]

>     print *, diff(i, append=[21]) !! [0, 1, 1, 2, 3, 5, 8]

Reviewer 3: Are the comments on these lines used by FORD in any way? If not, let's just use a single "!" for comments in examples. Otherwise, newcomers to the language may be confused why are two !! needed here.

    
end program demo_math_all_close
```

### `diff`

#### Description

Computes differences between adjacent elements of an array.

#### Syntax

> For rank-1 array

Reviewer: ```suggestion
For a rank-1 array
```

```fortran
y = [[stdlib_math(module):diff(interface)]](x [, n, prepend, append])
```

> and for rank-2 array

Reviewer: ```suggestion
and for a rank-2 array
```

```fortran
y = [[stdlib_math(module):diff(interface)]](x [, n, dim, prepend, append])
```

#### Status

Experimental.

#### Class

Pure function.

#### Arguments

> Note: The `x`, `prepend` and `append` arguments must have the same `type`, `kind` and `rank`.

Reviewer: I would suggest to move this note at the end of this section (with the other notes).

`x`: Shall be a `real/integer` and `rank-1/rank-2` array.
This argument is `intent(in)`.

`n`: Shall be an `integer` scalar.
This argument is `intent(in)` and `optional`, which is `1` by default.
It represents to calculate the n-th order difference.

`dim`: Shall be an `integer` scalar.
This argument is `intent(in)` and `optional`, which is `1` by default.
It gives the dimension of the input array along which the difference is calculated, between `1` and `rank(x)`.

`prepend`: Shall be a `real/integer` and `rank-1/rank-2` array, which is no value by default.
This argument is `intent(in)` and `optional`.

`append`: Shall be a `real/integer` and `rank-1/rank-2` array, which is no value by default.
This argument is `intent(in)` and `optional`.

Note: 
- If the value of `n` is less than or equal to `0`, the return value of `y` is `x`. (Not recommended)
- If the value of `dim` is not equal to `1` or `2`,
`1` will be used by the internal process of `diff`. (Not recommended)

#### Result value

> Note: That `y` generally has one fewer element than `x`.
> 
> Returns a `real/integer` and `rank-1/rank-2` array.

Reviewer: ```suggestion
Returns a `real/integer` and `rank-1/rank-2` array. When both `prepend` and `append ` are not present, the result `y` has one fewer element than `x` alongside the dimension `dim`.
```

    
end program demo_math_all_close
```

### `diff`

#### Description

Computes differences between adjacent elements of an array.

#### Syntax

For a rank-1 array
```fortran
y = [[stdlib_math(module):diff(interface)]](x [, n, prepend, append])
```
and for a rank-2 array
```fortran
y = [[stdlib_math(module):diff(interface)]](x [, n, dim, prepend, append])
```

#### Status

Experimental.

#### Class

Pure function.

#### Arguments

`x`: The array to take a difference of.
Shall be a `real/integer` and `rank-1/rank-2` array.
This argument is `intent(in)`.

> `n`: The order of finite differencing to use.
> Shall be an `integer` scalar.
> This argument is `intent(in)` and `optional`, which is `1` by default.

Reviewer 3: I apologize, now I realize that this wording could be confused for the accuracy order of the finite difference scheme. How about:

```suggestion
`n`: How many times to iteratively calculate the difference.
Shall be an `integer` scalar.
This argument is `intent(in)` and `optional`, and has value of `1` by default.
```

Author: Thanks, I feel good, now updated.
